library/core/src/slice/mod.rs

```diff
@@ -4851,20 +4851,30 @@ impl<T, const N: usize> SlicePattern for [T; N] {
 }
 
 /// This checks every index against each other, and against `len`.
-///
-/// This will do `binomial(N + 1, 2) = N * (N + 1) / 2 = 0, 1, 3, 6, 10, ..`
-/// comparison operations.
 fn get_many_check_valid<const N: usize>(indices: &[usize; N], len: usize) -> bool {
-    // NB: The optimizer should inline the loops into a sequence
-    // of instructions without additional branching.
-    let mut valid = true;
-    for (i, &idx) in indices.iter().enumerate() {
-        valid &= idx < len;
-        for &idx2 in &indices[..i] {
-            valid &= idx != idx2;
+    // Based on benchmarks, it is faster to sort starting with 9 indices.
+    if N >= 9 {
+        let mut sorted_indices = *indices;
```

Comment:
The commit message states:

> We only need an immutable access to the indices array, and it we take it owned we may force a copy for big arrays. For small arrays, LLVM will fully inline and unroll the body anyway, so this doesn't matter.

So it seems any gains there are unfortunately immediately lost for anything N >= 9. Depending on the exact performance it may be worth checking `indices.is_sorted()` before duplicating and sorting - do you have the results of those benchmarks?

Reply:
This is not true, as we need both the unsorted and sorted array. So if we take by value we would copy twice.

About checking `is_sorted()`, I can do that, but is it worth it? How common is sorted slice that you don't know is sorted? After all, if you know it's sorted you can use a guaranteed O(N) API (not in std) - and I don't know how common is this alone, yet alone by accidence.

However you made me realize that my assembly inspection was wrong - LLVM did copy the array, but only once. So it seems it is good enough to know when the copy can be elided. I wonder if that means we can trust it?

Reply:
I meant that if copying is showing up to be significant on benchmarks (enough to justify the by-val to by-ref change), then maybe we should try to avoid the copy whenever possible. But LLVM is usually pretty good about figuring out when it makes sense to pass larger types as a pointer, so I am curious how different the benchmark results turned out.

The `DisjointIndices` type would make this a lot cleaner since it gives a way to move copying and sorting out of a potentially hot loop.

Reply:
I didn't benchmark passing by ref/value, all I benchmarked was sorting versus O(N^2) check. Without a concrete example where LLVM does not elide the copy, of course benchmarks won't show any difference.

> The DisjointIndices type would make this a lot cleaner since it gives a way to move copying and sorting out of a potentially hot loop.

This is the "reusing indices" benefit, but as I said, I doubt how useful this is.